File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/PermissionManager.ios.kt
package org.example.kmpsamples.presentation.permissions

import org.example.kmpsamples.presentation.permissions.dtos.PermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType
import org.example.kmpsamples.presentation.permissions.dtos.PermissionType
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformMultiplePermissions
import org.example.kmpsamples.presentation.permissions.systemPermissionHandlers.SystemPermissionHandlerLocatorInterface
import platform.Foundation.NSURL
import platform.UIKit.UIApplication
import platform.UIKit.UIApplicationOpenSettingsURLString

actual interface PermissionManagerInterface {
    actual fun onPermissionState(onPermissionState: ((PermissionStateType) -> Unit)?)
    actual fun checkPermissions(permissionTypes: List<PlatformPermissionType>): List<PermissionState>
    actual fun requestPermission(permissionType: PlatformPermissionType)
    actual fun createPlatformSpecificPermissions(permissionTypes: List<PermissionType>): List<PlatformPermissionType>
    actual fun openGrantPermissionsScreen()
}

class IOSPermissionManager(private val systemPermissionHandlerLocator: SystemPermissionHandlerLocatorInterface) :
    PermissionManagerInterface {

    private var onPermissionState: ((PermissionStateType) -> Unit)? = null

    override fun onPermissionState(onPermissionState: ((PermissionStateType) -> Unit)?) {
        this.onPermissionState = onPermissionState
    }

    override fun createPlatformSpecificPermissions(permissionTypes: List<PermissionType>): List<PlatformPermissionType> {
        val platformPermissionTypes = permissionTypes.map { PlatformPermissionType.create(it) }
        return PlatformPermissionMapper.toPlatformSinglePermissions(platformPermissionTypes)
            .groupBy { systemPermissionHandlerLocator.getPermissionHandler(it) }
            .entries.map { entry ->
                PlatformMultiplePermissions(entry.value.toList())
            }
    }

    override fun checkPermissions(permissionTypes: List<PlatformPermissionType>): List<PermissionState> {
        return PermissionStateMapper.toPermissionStates(
            resolvePermissionStates(permissionTypes)
        )
    }

    override fun requestPermission(permissionType: PlatformPermissionType) {

        val onPermissionState = this.onPermissionState ?: return
        val requestedPermissionState = resolvePermissionStates(listOf(permissionType)).first()
        if (requestedPermissionState.isGranted) {
            onPermissionState(requestedPermissionState)
        } else {
            systemPermissionHandlerLocator.getPermissionHandler(permissionType)
                .requestPermission { resolvedPermissionState ->
                    onPermissionState(resolvedPermissionState)
                }
        }
    }

    override fun openGrantPermissionsScreen() {
        val url = NSURL.URLWithString(UIApplicationOpenSettingsURLString) ?: return
        UIApplication.sharedApplication.openURL(url, options = emptyMap<Any?, Any>()) { success ->
            println("Opened app settings: $success")
        }
    }

    private fun resolvePermissionStates(
        permissionTypes: List<PlatformPermissionType>
    ): List<PermissionStateType> {
        return permissionTypes.map {
            systemPermissionHandlerLocator.getPermissionHandler(it).resolvePermissionState()
        }
    }
}

File: iosApp/iosApp/Permissions/PermissionManager.ios.swift
import Foundation
import UIKit
import ComposeApp

class IOSPermissionManager: PermissionManagerInterface {
    private let systemPermissionHandlerLocator: SystemPermissionHandlerLocatorInterface
    private var onPermissionState: ((PermissionStateType) -> Void)?
    
    init(systemPermissionHandlerLocator: SystemPermissionHandlerLocatorInterface) {
        self.systemPermissionHandlerLocator = systemPermissionHandlerLocator
    }
    
    func onPermissionState(onPermissionState: ((PermissionStateType) -> Void)?) {
        self.onPermissionState = onPermissionState
    }
    
    func createPlatformSpecificPermissions(permissionTypes: [PermissionType]) -> [PlatformPermissionType] {
        let platformPermissionTypes: [PlatformPermissionType] = permissionTypes.map {
            PlatformPermissionTypeCompanion.shared.create(permissionType: $0) }
        let singlePlatformPermissions = PlatformPermissionMapper.toPlatformSinglePermissions(platformPermissionTypes)
        let groupedPermissions = Dictionary(grouping: singlePlatformPermissions
                                            
        ){ key in
            SystemPermissionHandlerWrapper(handler: systemPermissionHandlerLocator.getPermissionHandler(permissionType: key))
        }
        
        let worstCostValue = groupedPermissions.count + 1
        let entries = groupedPermissions.sorted(by: {left, right in
            var leftCostValue = worstCostValue
            if let el = left.value.first, let index = singlePlatformPermissions.firstIndex(of: el) {
                leftCostValue = index
            }
            
            var rightCostValue = worstCostValue
            if let el = right.value.first, let index = singlePlatformPermissions.firstIndex(of: el) {
                rightCostValue = index
            }
            
            return leftCostValue <= rightCostValue
        })
        
        return entries.map { entry in
            PlatformPermissionTypePlatformMultiplePermissions(permissions: Array(entry.value))
        }
    }
    
    func checkPermissions(permissionTypes: [PlatformPermissionType]) -> [PermissionState] {
        return PermissionStateMapper.shared.toPermissionStates(
            permissionStateTypes:
                resolvePermissionStates(
                    permissionTypes: permissionTypes
                )
        )
    }
    
    func requestPermission(permissionType: PlatformPermissionType) {
        guard let onPermissionState = self.onPermissionState else { return }
        let requestedPermissionState = resolvePermissionStates(permissionTypes: [permissionType]).first!
        
        if requestedPermissionState.isGranted == true {
            onPermissionState(requestedPermissionState)
        } else {
            systemPermissionHandlerLocator.getPermissionHandler(permissionType: permissionType)
                .requestPermission { resolvedPermissionState in
                    DispatchQueue.main.async {
                        onPermissionState(resolvedPermissionState)
                    }
                }
        }
    }
    
    func openGrantPermissionsScreen() {
        guard let url = URL(string: UIApplication.openSettingsURLString),
              UIApplication.shared.canOpenURL(url) else { return }
        UIApplication.shared.open(url, options: [:]) { success in
            print("Opened app settings: \(success)")
        }
    }
    
    private func resolvePermissionStates(permissionTypes: [PlatformPermissionType]) -> [PermissionStateType] {
        return permissionTypes.map {
            systemPermissionHandlerLocator.getPermissionHandler(permissionType: $0).resolvePermissionState()
        }
    }
}
